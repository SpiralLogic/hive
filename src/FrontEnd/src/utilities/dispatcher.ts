import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
