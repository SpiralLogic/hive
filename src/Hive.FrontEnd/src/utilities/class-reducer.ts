import { useReducer } from 'preact/hooks';

const classReducer = (
  initialClasses: string,
  action: { type: 'add' | 'remove'; classes: string[] }
): string => {
  const classList = new Set(initialClasses.split(' '));
  switch (action.type) {
    case `add`:
      for (const c of action.classes) classList.add(c);
      break;

    case 'remove':
    default:
      for (const c of action.classes) classList.delete(c);
      break;
  }
  classList.delete('');
  return [...classList].join(' ');
};

export const useClassReducer = (
  initialClasses: string
): [string, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
