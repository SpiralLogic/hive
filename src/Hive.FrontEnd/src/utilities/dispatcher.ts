import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const dispatcher = useHiveDispatcher();
    return dispatcher.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
