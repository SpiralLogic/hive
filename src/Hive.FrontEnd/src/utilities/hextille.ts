import { Cell, GameState, HexCoordinates, Player, PlayerId, Tile } from '../domain';

const getAllTiles = (...parents: Array<Array<Player | Cell>>): Array<Tile> =>
  parents.flatMap((g) => g.flatMap((p) => p.tiles));

const getAllPlayerTiles = (playerId: PlayerId, ...parents: Array<Array<Player | Cell>>) =>
  getAllTiles(...parents).filter((t) => t.playerId !== playerId);

export const cellKey = ({ q, r }: HexCoordinates) => `${q}-${r}`;

export const removeOtherPlayerMoves = (
  playerId: number,
  { players, cells }: Pick<GameState, 'players' | 'cells'>
): void => {
  for (const t of getAllPlayerTiles(playerId, players, cells)) t.moves.splice(0, t.moves.length);
};
