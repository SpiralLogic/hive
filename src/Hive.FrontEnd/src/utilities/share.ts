import { PlayerId } from '../domain';

const fallbackCopyTextToClipboard = (text: string) => {
  const currentFocus = document.activeElement as HTMLElement;
  const textArea = document.createElement('textarea');
  textArea.value = text;

  textArea.style.top = '0';
  textArea.style.left = '0';
  textArea.style.position = 'fixed';

  document.body.append(textArea);
  textArea.focus();
  textArea.select();
  let copied = false;
  try {
    document.execCommand('copy');
    currentFocus.focus();
    copied = true;
  } catch {
    copied = false;
  }
  textArea.remove();
  return copied;
};

export const getShareUrl = (currentPlayer: PlayerId) => {
  const parts = window.location.href.split('/');
  parts[parts.length - 1] = currentPlayer === 1 ? '0' : '1';
  return parts.join('/');
};

export const shareGame = async (currentPlayer: PlayerId) => {
  const url = getShareUrl(currentPlayer);
  const opponentGame = {
    title: 'Hive board game',
    text: 'Click this link to join me in a game of the Hive!',
    url,
  };
  try {
    await navigator.share(opponentGame);
    return false;
  } catch {
    try {
      await navigator.clipboard.writeText(opponentGame.url);
      return true;
    } catch {
      return fallbackCopyTextToClipboard(opponentGame.url);
    }
  }
};
