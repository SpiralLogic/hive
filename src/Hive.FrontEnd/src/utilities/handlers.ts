import { GameState, Tile } from '../domain';
import { EngineMove, OpponentConnectedHandler, OpponentSelectionHandler } from '../domain/engine';
import { AiAction, HiveDispatcher, MoveEvent, TileEvent } from '../services';
import { dispatchHiveEvent } from './dispatcher';

export function handleDragOver(event_: { preventDefault: () => void }): boolean {
  event_.preventDefault();
  return false;
}

export function handleDrop(event_: { preventDefault: () => void }): boolean {
  event_.preventDefault();
  return false;
}

export const isEnterOrSpace = (event: KeyboardEvent): boolean => event.key === 'Enter' || event.key === ' ';

const focusNext = (target: HTMLElement, direction: -1 | 1) => {
  const allTabbable = [...document.querySelectorAll('*[tabindex]:not(.name)')];
  let index = allTabbable.indexOf(target);
  if (index + direction < 0) index = allTabbable.length;
  (allTabbable[(index + direction) % allTabbable.length] as HTMLElement).focus();
};

export const handleKeyboardNav = (error: Pick<KeyboardEvent, 'key' | 'target'>): boolean => {
  if (error.target instanceof HTMLElement) {
    switch (error.key) {
      case 'ArrowDown':
      case 'ArrowRight':
        focusNext(error.target, 1);
        return true;

      case 'ArrowUp':
      case 'ArrowLeft':
        focusNext(error.target, -1);
        return true;

      default:
        return false;
    }
  }
  return false;
};

export const opponentSelectionHandler: OpponentSelectionHandler = (type, tile) => {
  switch (type) {
    case 'select':
      dispatchHiveEvent({ type: 'tileSelect', tile });
      break;

    case 'deselect':
      dispatchHiveEvent({ type: 'tileClear' });
      break;

    default:
      break;
  }
};
export const opponentConnectedHandler: OpponentConnectedHandler = (type) => {
  switch (type) {
    case 'connect':
      dispatchHiveEvent({ type: 'opponentConnected' });
      dispatchHiveEvent<AiAction>({ type: 'toggleAi', newState: false });
      break;

    case 'disconnect':
      dispatchHiveEvent({ type: 'opponentDisconnected' });
      break;

    default:
      break;
  }
};

export const addServerHandlers = (
  sendSelection: (type: 'select' | 'deselect', tile: Tile) => void,
  gameState: GameState,
  updateGameState: (value: GameState) => void,
  move: EngineMove,
  useAi = false,
  hiveDispatcher: HiveDispatcher
) => {
  const selectionChangeHandler = (event: TileEvent) =>
    !event.fromEvent && sendSelection('select', event.tile);
  const deselectionChangeHandler = (event: TileEvent) =>
    !event.fromEvent && sendSelection('deselect', event.tile);
  const moveHandler = async (event: MoveEvent) => {
    const state = await move(gameState.gameId, event.move, useAi);
    return updateGameState(state);
  };

  hiveDispatcher.add<MoveEvent>('move', moveHandler);
  hiveDispatcher.add<TileEvent>('tileSelected', selectionChangeHandler);
  hiveDispatcher.add<TileEvent>('tileDeselected', deselectionChangeHandler);
  const removeAi = hiveDispatcher.add<AiAction>('toggleAi', ({ newState }) => (useAi = newState));
  return () => {
    hiveDispatcher.remove<MoveEvent>('move', moveHandler);
    hiveDispatcher.remove<TileEvent>('tileSelected', selectionChangeHandler);
    hiveDispatcher.remove<TileEvent>('tileDeselected', deselectionChangeHandler);
    removeAi();
  };
};
