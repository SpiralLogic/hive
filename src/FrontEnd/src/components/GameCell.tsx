import { FunctionComponent, h } from 'preact';
import { HexCoordinates, Tile as TileType } from '../domain';
import { MoveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { addHiveDispatchListener, dispatchHiveEvent, useClassReducer } from '../utilities/hooks';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Hexagon from './Hexagon';

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  const isValidMove = (validMoves: HexCoordinates[]) =>
    validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);
  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'remove', classes: ['no-drop'] });
    }
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(event.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves)) {
      dispatchHiveEvent<MoveEvent>({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (!handleKeyboardNav(e) && isEnterOrSpace(e)) return handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
