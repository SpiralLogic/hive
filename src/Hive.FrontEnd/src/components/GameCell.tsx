import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { HexCoordinates, Tile as TileType } from '../domain';
import { TileEvent } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useClassReducer } from '../utilities/class-reducer';
import Hexagon from './Hexagon';
const isValidMove = (validMoves: HexCoordinates[], coords: HexCoordinates) =>
  validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', () => {
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop', 'no-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (intent) => {
    if (!isValidMove(intent.tile.moves, coords)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(intent.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves, coords)) {
      dispatchHiveEvent({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves, coords))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves, coords) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
