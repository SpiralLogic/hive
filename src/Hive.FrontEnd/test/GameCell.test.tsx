import { fireEvent, render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { getHiveDispatcher } from '../src/utilities/dispatcher';
import GameCell from '../src/components/GameCell';
import { simulateEvent } from './test-helpers';
import {
  createCellCanDrop,
  createCellNoDrop,
  createCellWithNoTile,
  createCellWithTile,
  createCellWithTileAndDrop,
  createCellWithTileNoDrop,
  createDispatcher,
  emitHiveEvent,
  moveTileSpy,
} from './fixtures/game-cell.fixtures';

describe('<GameCell>', () => {
  it(`top tile is rendered when tiles are all empty`, () => {
    render(<GameCell {...createCellWithTile()} />);
    expect(screen.getByTitle(/fly/)).toBeInTheDocument();
  });

  describe('drag and drop', () => {
    it('dragover allows drop', () => {
      render(<GameCell {...createCellWithTile()} />);
      const preventDefault = simulateEvent(screen.getByRole('cell'), 'dragover');
      expect(preventDefault).toHaveBeenCalledWith();
    });

    it('cell is available on drag start', () => {
      render(<GameCell {...createCellWithTileAndDrop()} />);
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');

      for (const c of screen.getAllByRole('cell')) expect(c).toHaveClass('can-drop');
    });

    it('available cell is active on tile drag enter', () => {
      render(<GameCell {...createCellWithTileAndDrop()} />);
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');

      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      for (const c of screen.getAllByRole('cell')) expect(c).toHaveClass('active');
    });

    it('available cell is no longer active on tile drag leave', () => {
      render(<GameCell {...createCellWithTileAndDrop()} />);
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      for (const c of screen.getAllByRole('cell')) fireEvent.dragLeave(c);
      for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('active');
    });

    it('move calls moves tile when cell is valid and active', () => {
      const moveEvents = createDispatcher();
      render(<GameCell {...createCellWithTileAndDrop()} />);
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      emitHiveEvent('tileDropped');

      expect(moveEvents).toEqual(
        expect.arrayContaining([
          {
            type: 'move',
            move: { tileId: 2, coords: { q: 0, r: 0 } },
          },
          {
            type: 'move',
            move: { tileId: 2, coords: { q: 2, r: 2 } },
          },
        ])
      );
    });

    it(`drop doesn't call move tile when cell doesn't allow drop`, () => {
      jest.spyOn(getHiveDispatcher(), 'dispatch');
      render(<GameCell {...createCellWithTile()} />);
      render(<GameCell {...createCellNoDrop()} />);
      emitHiveEvent('tileSelected');
      emitHiveEvent('tileDropped');

      expect(getHiveDispatcher().dispatch).not.toHaveBeenCalledWith({ type: 'move' });
    });

    it(`invalid cells don't call move tile on drop`, () => {
      render(<GameCell {...createCellWithTileNoDrop()} />);
      render(<GameCell {...createCellNoDrop()} />);
      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      expect(moveTileSpy).not.toHaveBeenCalledWith();
    });

    it('active classes are removed on drag leave', () => {
      createCellWithTile();
      createCellWithNoTile();
      render(<GameCell {...createCellWithTileNoDrop()} />);
      render(<GameCell {...createCellNoDrop()} />);
      emitHiveEvent('tileSelected');

      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      for (const c of screen.getAllByRole('cell')) fireEvent.dragLeave(c);
      for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('active');
    });

    it('active and can-drop classes are removed on drop', () => {
      render(<GameCell {...createCellWithTile()} />);
      createCellWithNoTile();
      render(<GameCell {...createCellWithTileNoDrop()} />);
      render(<GameCell {...createCellNoDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
      emitHiveEvent('tileDropped');
      emitHiveEvent('tileDeselected');

      for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('active');
      for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('can-drop');
    });

    it(`occupied cell with no active tile doesn't stop event propagation'`, () => {
      jest.spyOn(getHiveDispatcher(), 'dispatch');
      render(<GameCell {...createCellWithTileAndDrop()} />);
      for (const c of screen.getAllByRole('cell')) userEvent.click(c);

      expect(getHiveDispatcher().dispatch).not.toHaveBeenCalledWith();
    });

    it(`cell click with active tile makes a move`, async () => {
      const moveEvents = createDispatcher();
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of await screen.findAllByRole('cell')) userEvent.click(c);

      expect(moveEvents).toEqual(
        expect.arrayContaining([
          {
            type: 'move',
            move: { tileId: 2, coords: { q: 0, r: 0 } },
          },
        ])
      );
    });

    it(`cell click with no active tile shouldn't move`, () => {
      jest.spyOn(getHiveDispatcher(), 'dispatch');
      render(<GameCell {...createCellNoDrop()} />);
      for (const c of screen.getAllByRole('cell')) userEvent.click(c);

      expect(getHiveDispatcher().dispatch).not.toHaveBeenCalledWith();
    });

    it(`cell click with invalid tile shouldn't move`, async () => {
      jest.spyOn(getHiveDispatcher(), 'dispatch');
      render(<GameCell {...createCellNoDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of await screen.findAllByRole('cell')) userEvent.click(c);

      expect(getHiveDispatcher().dispatch).not.toHaveBeenCalledWith();
    });

    it(`enter fires emit event on keydown enter`, async () => {
      const moveEvents = createDispatcher();
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of await screen.findAllByRole('cell')) userEvent.type(c, '{enter}');

      expect(moveEvents).toEqual(expect.arrayContaining([expect.objectContaining({ type: 'move' })]));
    });

    it(`space fires emit event on keydown enter`, async () => {
      const moveEvents = createDispatcher();
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      for (const c of await screen.findAllByRole('cell')) userEvent.type(c, ' ');

      expect(moveEvents).toEqual(expect.arrayContaining([expect.objectContaining({ type: 'move' })]));
    });

    it(`other keys dont emit tile start event`, async () => {
      jest.spyOn(getHiveDispatcher(), 'dispatch');
      render(<GameCell {...createCellCanDrop()} />);
      emitHiveEvent('tileSelected');
      userEvent.type(await screen.findByRole('cell'), 'j');

      expect(getHiveDispatcher().dispatch).not.toHaveBeenCalledWith();
    });

    it('cell with tile matches current snapshot', () => {
      render(<GameCell {...createCellWithTile()} />);
      expect(screen.getByRole('cell')).toMatchSnapshot();
    });

    it('cell with no tile matches current snapshot', () => {
      render(<GameCell {...createCellWithNoTile()} />);
      expect(screen.getByRole('cell')).toMatchSnapshot();
    });
  });
});
