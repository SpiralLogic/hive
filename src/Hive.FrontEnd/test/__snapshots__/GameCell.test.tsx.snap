// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cell Tests drag and drop cell with no tile matches current snapshot 1`] = `
<div
  role="cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
</div>
`;

exports[`cell Tests drag and drop cell with tile matches current snapshot 1`] = `
<div
  role="cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
  <div
    class="player1 hex tile fly"
    title="Player-1 fly"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
      <use
        class="creature"
        href="#fly"
      />
    </svg>
  </div>
</div>
`;
